### PR TITLE
Feature/Add LLMs Integration For Oracle Cloud Infrastructure(OCI) Data Science Model Deployment Endpoint

### PR DESCRIPTION
**Pull request by mingkang111**
- **Description:** 
    - [OCI Data Science](https://docs.oracle.com/en-us/iaas/data-science/using/home.htm) is a fully managed and serverless platform for data science teams to build, train, and manage machine learning models in the Oracle Cloud Infrastructure. This PR add integration for using LangChain with an LLM hosted on a [OCI Data Science Model Deployment](https://docs.oracle.com/en-us/iaas/data-science/using/model-dep-about.htm). To authenticate, [oracle-ads](https://accelerated-data-science.readthedocs.io/en/latest/user_guide/cli/authentication.html) has been used to automatically load credentials for invoking endpoint.
- **Issue:** None
- **Dependencies:** `oracle-ads`
- **Tag maintainer:** @baskaryan
- **Twitter handle:** None

### CODE REVIEW

**Comment by mingkang111** [comment]
Hi @ baskaryan, I have prepared test evidences page for this PR. It includes test results with endpoint, formate/lint/spell check as informed in contribution guideline. Feel free to let me know if you want to check with it. Thanks for helping with reviewing!

File: libs/langchain/langchain/llms/oci_data_science_model_deployment_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+        response = self.send_request(
+            data=body, endpoint=self.endpoint, **kwargs.pop("requests_kwargs", {})
+        )
+        completion = self._process_response(response)
+        logger.info(f"LLM API Completion:\n{completion}")
+        return completion
+
+    def send_request(
+        self,
+        data: Any,
```
Comment (mingkang111) [inline libs/langchain/langchain/llms/oci_data_science_model_deployment_endpoint.py:155]: Screenshot 2023-12-05 at 9 58 39 AM
Looks like it cannot recognize Any as type? cc: @baskaryan

Comment (hwchase17) [reply libs/langchain/langchain/llms/oci_data_science_model_deployment_endpoint.py:155]: you havent typed **kwargs, that should be `**kwargs: Any`

Comment (mingkang111) [reply libs/langchain/langchain/llms/oci_data_science_model_deployment_endpoint.py:155]: Thank you @hwchase17 ! I fixed it.

**Comment by mingkang111** [comment]
Hi @hwchase17 @efriis , thanks for helping with reviewing! May I know is there any change I need to make for this PR? Normally what is the timeline for merging the PR and release feature ( I saw release is ad-hoc from [here](https://github.com/langchain-ai/langchain/blob/master/.github/CONTRIBUTING.md#-release-process), may I know is there any proposed date for the next release) ?  Thanks again for your help!

**Comment by efriis** [comment]
Hey @mingkang111 ! Was having some trouble with the OCI signup flow and was unable to create an API key. If that's something you can help with, let me know! Otherwise I'll try again next week. 

We're in the middle of a big refactor that'll make our packages more modular (here's some [context](https://github.com/langchain-ai/langchain/discussions/14243)). because of that we likely won't be landing or releasing any new features until later next week.

**Comment by mingkang111** [comment]
Hi @efriis ! Definitely, I can help with the OCI signup flow. Here is the doc about setting API Keys: [link](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/apisigningkey.htm). Feel free to let me know if there any issues you see when following this guideline. We can also meet through zoom.

Thanks for sharing the information!

**Comment by streamnsight** [comment]
@efriis @hwchase17 
Thanks for this great contribution.
I'm with OCI Data and AI team, and I have one concern with the way this is organized, especially the fact that OCI has an official Gen AI Service and is working on contributing the classes for LLM and Embeddings in Langchain.
So you beat us to the punch there, which is appreciated, however you are using the generic model deployment endpoint, not the specific Gen AI endpoints. 
So in terms of the naming conventions, I would refrain from suffixing this work with xxxLLM as these are not LLM endpoints, and that would make it confusing to OCI users.
I think it should just be called BaseModelDeployment, XYZModelDeployment etc... and should support **kwargs so any param could be passed and made generic, and refrain from defining model specific classes. 
If you want to showcase using Llama2 deployed with this endpoint and using it for different use cases, that sounds great, however that should be part of a guide on how to use the ModelDeployment class / endpoints, as it seems like defining specific classes for this purpose and only exposing that in the docs greatly narrows the purpose of the ModelDeployment endpoint and will potentially clash with the official Gen AI service in the future.
Happy to discuss.

**Comment by streamnsight** [comment]
@efriis @mingkang111 @hwchase17 
I guess I missed the fact that @mingkang111 is also with Oracle, so my apologies for the confusion.
I want to clarify that I am not against this PR, I am just suggesting that if the base class was a ModelDeployment class (and not LLM specific), that would allow other users to deploy their own (generic or specialized) models using that base.
Then specific TGI and VLLM LLM classes based on ModelDeployment sound great for those who want to focus on using those specific models deployments frameworks

**Comment by mingkang111** [comment]
Hi @efriis , I just resolved the merge conflicts with the master branch. Can I get your review again? Thanks a lot for your help!

**Comment by efriis** [comment]
Thanks! Looks like it's still failing some CI checks if you want to address those.

@streamnsight would appreciate a 👍 or comments from you as well if your team is planning on contributing too! Any structural/naming changes get much more difficult after the initial merge because we maintain backwards compatibility.

**Comment by streamnsight** [comment]
@efriis 
looks good to me. Thanks

**Comment by mingkang111** [comment]
> @efriis @mingkang111 @hwchase17 I guess I missed the fact that @mingkang111 is also with Oracle, so my apologies for the confusion. I want to clarify that I am not against this PR, I am just suggesting that if the base class was a ModelDeployment class (and not LLM specific), that would allow other users to deploy their own (generic or specialized) models using that base. Then specific TGI and VLLM LLM classes based on ModelDeployment sound great for those who want to focus on using those specific models deployments frameworks

Hi @streamnsight , thanks for your suggestion. We currently offering OCIModelDeploymentTGI/OCIModelDeploymentVLLM as we provided these two types of deployment (please feel free to explore [here](https://github.com/oracle-samples/oci-data-science-ai-samples/tree/main/model-deployment/containers/llama2#overview)), In our ongoing commitment to enhance user experience, we are excited to integrate these deployments with LangChain, offering our users a seamless and efficient interface. Our dedicated team is in the process of developing more generic deployments, and we will have another PR to accommodate more generic deployments in the future. We are enthusiastic about these developments and believe they will significantly benefit our users.

**Comment by mingkang111** [comment]
Hi @efriis , thank you for your timely reminder. We are aware of the potential changes on the horizon and wish to assure you that these future updates are designed with great care. Our aim is to ensure they do not impact existing structures or naming conventions, thereby avoiding any backward compatibility issues.  

I have addressed the issues with the failing CI checks and adapted our project to align with the recent changes in LangChain, the transition to `langchain-community`. I would greatly appreciate it if you could review these updates. Your feedback is invaluable to the continuous improvement of our work. 

Thanks again for your time!

**Comment by mingkang111** [comment]
Hello @efriis , I noticed there is a release happening soon: https://github.com/langchain-ai/langchain/pull/14867. May I know is that possible to add this change into the release as well? Thank you!

**Comment by efriis** [comment]
Hey @mingkang111 ! We actually already missed the community release in #14864 (this will be in community).

I'm actually still running into the same error simply going through the account sign-up flow :(. Could you email me at erick@langchain.dev and we can figure out testing?

Screenshot 2023-12-18 at 1 29 59 PM

**Comment by mingkang111** [comment]
> Hey @mingkang111 ! We actually already missed the community release in #14864 (this will be in community).
> 
> I'm actually still running into the same error simply going through the account sign-up flow :(. Could you email me at [erick@langchain.dev](mailto:erick@langchain.dev) and we can figure out testing?
> 
> Screenshot 2023-12-18 at 1 29 59 PM

Absolutely! I will send you the demo soon! Thanks for your time!

May I know is there any community release before holiday?

**Comment by efriis** [comment]
Yes there will be more! We usually release a few times per week

**Comment by mingkang111** [comment]
> Yes there will be more! We usually release a few times per week

Thank you @efriis 

I have sent the demo to [erick@langchain.dev](mailto:erick@langchain.dev). Please don't hesitate to reach out if you have any questions or need further information!

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        body = self._construct_json_body(prompt, params)
+        logger.info(f"LLM API Request:\n{prompt}")
+        response = self.send_request(
+            data=body, endpoint=self.endpoint, **requests_kwargs
+        )
+        completion = self._process_response(response)
+        logger.info(f"LLM API Completion:\n{completion}")
+        return completion
+
+    def send_request(
```
Comment (efriis) [inline libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:155]: Is this necessary for clients to know about? Could we rename to not be public?
```suggestion
    def _send_request(
```

File: docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb
Comment (efriis) [inline docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb:1]: Also update to be from `langchain_community` on imports

Comment (mingkang111) [reply docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb:1]: Updated!

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py (hunk 2 above)
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Sure. Changed to private method. Thanks!

File: libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    @classmethod
+    def is_lc_serializable(cls) -> bool:
+        """This class can be serialized with default LangChain serialization."""
+        return True
+
+    @classmethod
+    def get_lc_namespace(cls) -> List[str]:
+        """Get the namespace of the langchain object."""
+        return ["langchain", "llms", "oci_data_science_model_deployment_endpoint"]
+
```
[the remark this answers is not in the document]
Comment (mingkang111) [reply libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py]: Removed.

File: libs/core/langchain_core/load/mapping.py
[the remark this answers is not in the document]
Comment (mingkang111) [reply libs/core/langchain_core/load/mapping.py:1]: Hi @efriis , Sure, I will revert this change. However, we do want to support serialization, is this mean we should create a separate PR to make the related change?  Thanks!

File: libs/langchain/langchain/llms/__init__.py
[the remark this answers is not in the document]
Comment (mingkang111) [reply libs/langchain/langchain/llms/__init__.py:1]: Reverted!

File: libs/community/tests/integration_tests/llms/test_oci_model_deployment_endpoint.py
[the remark this answers is not in the document]
Comment (mingkang111) [reply libs/community/tests/integration_tests/llms/test_oci_model_deployment_endpoint.py:1]: Moved the test to unit test.

File: libs/community/tests/integration_tests/llms/test_oci_model_deployment_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        responses.POST,
+        endpoint,
+        json={
+            "choices": [{"index": 0, "text": "This is a completion."}],
+        },
+        status=200,
+    )
+
+    llm = OCIModelDeploymentVLLM(endpoint=endpoint, model="my_model")
+    output = llm("This is a prompt.")
```
[the remark this answers is not in the document]
Comment (mingkang111) [reply libs/community/tests/integration_tests/llms/test_oci_model_deployment_endpoint.py]: Fixed! Thanks for the reminder!

File: libs/core/langchain_core/load/mapping.py
[the remark this answers is not in the document]
Comment (efriis) [reply libs/core/langchain_core/load/mapping.py:1]: Serialization is a separate feature here, focused on the LangSmith playground!

[the remark this answers is not in the document]
Comment (mingkang111) [reply libs/core/langchain_core/load/mapping.py:1]: Thanks @efriis ,  I will check LangSmith for the support!

**Review by efriis — APPROVED** [review]

**Comment by mingkang111** [comment]
Hi @efriis ! Thanks for helping to resolve extra dependency test! I noticed some conflicts with master branch, and fixed them. Here is the screenshot of running lint check (failed before) against libs/experimental : 
Screenshot 2023-12-19 at 9 27 51 PM
Reset my environment and test again:
Screenshot 2023-12-19 at 11 58 56 PM

**Pull request merged**